File: internal/installer/steps/openai_gateway.go
package steps

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"fmt"

	"github.com/SolsticeSauer/solsclaw_beta/internal/installer"
	"github.com/SolsticeSauer/solsclaw_beta/internal/secrets"
)

type OpenAIGateway struct{}

func (OpenAIGateway) ID() string    { return "openai-gateway" }
func (OpenAIGateway) Label() string { return "Enable OpenClaw OpenAI-compatible gateway" }

func (OpenAIGateway) ShouldRun(sc installer.StepContext) bool {
	return sc.Submission.OptionalFeatures.OpenAIGateway
}

func (s OpenAIGateway) Run(ctx context.Context, sc installer.StepContext) error {
	buf := make([]byte, 24)
	if _, err := rand.Read(buf); err != nil {
		return err
	}
	token := "oc_" + hex.EncodeToString(buf)

	ref, _, err := secrets.Set("gateway:openai-compat-token", token)
	if err != nil {
		return err
	}
	sc.Bus.Log(s.ID(), "info", fmt.Sprintf("Issued shared-secret token (%s).", ref))

	// Best-effort daemon-side toggle. The on-disk config we wrote earlier
	// already carries gateway.openaiCompat.enabled, so the daemon will pick
	// it up on its next config reload regardless.
	if err := streamCommand(ctx, sc, s.ID(), "openclaw", "gateway", "set-openai-compat", "--enable", "--token-ref", ref); err != nil {
		sc.Bus.Log(s.ID(), "warn", "Daemon-side toggle skipped: "+err.Error())
	}
	return nil
}

File: packages/installer-server/src/lib/steps/openaiGateway.ts
import crypto from 'node:crypto';
import { execa } from 'execa';
import { setSecret } from '../secretsStore.js';
import type { InstallerStep } from './index.js';

export const openaiGatewayStep: InstallerStep = {
  id: 'openai-gateway',
  label: 'Enable OpenClaw OpenAI-compatible gateway endpoint',
  shouldRun: ({ submission }) => submission.optionalFeatures.openaiGateway,
  async run({ bus }) {
    const token = `oc_${crypto.randomBytes(24).toString('hex')}`;
    const { ref } = await setSecret('gateway:openai-compat-token', token);
    bus.log(openaiGatewayStep.id, 'info', `Issued shared-secret token (${ref}).`);

    // Tell OpenClaw to enable the OpenAI-compatible surface and reload its
    // config. The CLI subcommand documented in the gateway runbook keeps the
    // installer free of direct mutation against an already-running daemon.
    await execa('openclaw', ['gateway', 'set-openai-compat', '--enable', '--token-ref', ref], { stdio: 'inherit' }).catch(
      (err: unknown) => {
        // If the subcommand differs in this OpenClaw version, leave a hint —
        // the config we wrote earlier already contains gateway.openaiCompat.
        const message = err instanceof Error ? err.message : String(err);
        bus.log(openaiGatewayStep.id, 'warn', `Daemon-side toggle skipped: ${message}`);
      },
    );
  },
};
